[JENKINS-26097] Adjust label expression auto-completion and validation

Zastai opened the pull request:
See [JENKINS-26097](https://issues.jenkins-ci.org/browse/JENKINS-26097).

This moves label expression auto-completion and validation from `AbstractProject` to `LabelExpression`.

Auto-Completion:
- `AbstractProject.AutoCompleteSeeder` was moved to `LabelExpression`
  - test class correspondingly moved to the relevant package
- added static `LabelExpression.autoComplete()` method
  - does what the non-static `AbstractProject.doAutoCompleteLabel()` used to do; that now calls the new static method

Validation:
- added `LabelExpression.LabelValidator`, which is like `AbstractProject.LabelValidator` except it takes a `Job` instead of an `AbstractProject`
  - `AbstractProject.LabelValidator` is now deprecated
- added static `LabelExpression.validate(String, Job)` method, which replaces `AbstractProject.validateLabelExpression(String, AbstractProject)`
  - the latter is marked as deprecated and forwards to the former
  - any warnings/errors reported by validators are now aggregated (previously only the first one was shown)
- `LabelExpressionTest.formValidation()` uses the new static method
- associated messages were moved from `hudson.model.Messages.AbstractProject_xxx()` to
  `hudson.model.Messages.LabelExpression_xxx()`).

### Proposed changelog entries

* JENKINS-26097: `hudson.model.AbstractProject.LabelValidator` is now deprecated in favour of `jenkins.model.labels.LabelValidator`
  * but the old version also has a new `checkItem()` method to allow them to apply validation to non-Project items too
* JENKINS-26097: `hudson.model.AbstractProject.DescriptorImpl.validateLabelExpression()` is now deprecated in favour of `LabelExpression.validate()` (which takes any kind of `Item` object instead of only `AbstractProject`s)
  * this will now aggregate all warnings and errors reported by `LabelValidator`s (old and new)
* JENKINS-26097: auto-completion for labels is now available via `LabelExpression.autoComplete()`

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [X] Appropriate autotests or explanation to why this change has no tests
- [X] For dependency updates: links to external changelogs and, if possible, full diffs

Reference implementations:
* https://github.com/jenkinsci/matrix-project-plugin/pull/63
* https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/394
* https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/136

### Desired reviewers

@jglick 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

Zastai committed:
JENKINS-26097: Adjust label expression auto-completion and validation.

This moves label expression auto-completion and validation from AbstractProject
to LabelExpression.

Auto-Completion:
- AbstractProject.AutoCompleteSeerder was moved to LabelExpression
  - test class correspondingly moved to the relevant package
- added static LabelExpression.autoComplete() method
  - does what the non-static AbstractProject.doAutoCompleteLabel() used to do; that now calls the new static method

Validation:
- added LabelExpression.LabelValidator, which is like AbstractProject.LabelValidator except it takes a Job instead of an AbstractProject
  - AbstractProject.LabelValidator is now deprecated
- added static LabelExpression.validate(String, Job) method, which replaces AbstractProject.validateLabelExpression(String, AbstractProject)
  - the latter is marked as deprecated and forwards to the former
- LabelExpressionTest.formValidation() uses the new static method

Zastai committed:
Add auto-completion and validation for the label fields on tool installers.

Zastai committed:
Merge branch 'master' into JENKINS-26097

Zastai committed:
Use TODO, not FIXME for @since.

jglick commented on core/src/main/java/hudson/model/labels/LabelExpression.java:
```diff
@@ -210,4 +225,158 @@ public LabelOperatorPrecedence precedence() {
+     *
+     * @since TODO
+     */
+    public static abstract class LabelValidator implements ExtensionPoint {
```
Could be an `interface` I suppose.

Zastai replied:
Sure. I tried keeping it as close to the original as possible.

Zastai replied:
Have done this now.

jglick commented on core/src/main/java/hudson/model/labels/LabelExpression.java:
```diff
@@ -210,4 +225,158 @@ public LabelOperatorPrecedence precedence() {
+        if (job != null) {
+            if (job instanceof AbstractProject) { // Use any project-oriented label validators
+                final AbstractProject<?, ?> project = (AbstractProject<?,?>) job;
+                for (AbstractProject.LabelValidator v : j.getExtensionList(AbstractProject.LabelValidator.class)) {
```
or can use `ExtensionList.lookup`

Zastai replied:
I just kept the original code (other than wrapping it in a is-it-a-project check); I can look at using this other method instead.

jglick replied:
> I just kept the original code (other than wrapping it in a is-it-a-project check)

Fine, this was just FYI.

jglick commented on core/src/main/java/hudson/model/labels/LabelExpression.java:
```diff
@@ -210,4 +225,158 @@ public LabelOperatorPrecedence precedence() {
+         * @return The validation result.
+         */
+        @NonNull
+        public abstract FormValidation check(@NonNull Job<?, ?> job, @NonNull Label label);
```
Why not make `job` be `@CheckForNull` in case a validator can work without a job context?

Or relax the context type to `Item`?

Zastai replied:
I think the validator concept was added specifically for project-based validation (the main validation code already handles pure label expression validation). It was certainly only ever invoked when a project was specified.

But given that this is new API specifically to relax the Project-oriented existing API, I have no issue using Item instead. I lack the familiarity with the codebase to know what additional cases that would enable though.

jglick replied:
> I lack the familiarity with the codebase to know what additional cases that would enable though.

Mainly validation on `Folder`s, in case you have something that offers certain labels only in some subfolders. Something like https://github.com/jenkinsci/kubernetes-plugin/blob/d9cf6a810fce807fe35cd56ebe17ffb343c702f4/src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesFolderProperty/config.jelly#L5 perhaps. (CloudBees CI also has a proprietary example I think.)

Zastai replied:
Done - `Item` is now used instead of `Job`.

jglick reviewed (commented):
Looks reasonable from a quick glance. (Blocks of code moved around makes it hard to read the diff.) Would make sense to consume this from `workflow-durable-task-step-plugin` so that snippet generator on the `node` step can offer the same conveniences that you get in a freestyle project’s config screen.

Zastai committed:
Use Item as validation context for the new API (instead of Job).

In addition, the new LabelValidators are now applied to AbstractProjects too.

Zastai committed:
Make LabelExpression.LabelValidator an interface.

Zastai committed:
Fix a JavaDoc reference.

jglick reviewed (approved):
Would still recommend demonstrating usage in `workflow-durable-task-step-plugin`.

Zastai committed:
Merge branch 'master' into JENKINS-26097

Zastai commented:
I merged current master so the build produces an incremental; will look at creating a PR on `workflow-durable-task-step-plugin` using that incremental.

Zastai commented:
Looking at `workflow-durable-task-step-plugin` I am now confused. I always thought that `label` property was just informational text and was in no way related to the label expressions you have elsewhere (like tool installers and projects). So I'm not sure it makes sense adding the label expression validation and autocompletion there.

If there was a snippet generator for the `agent` part of a declarative pipeline (or the `node` for a scripted one), then sure, it makes sense there.

Zastai commented:
I had missed `ExecutorStep`; that does have an applicable field.

jglick commented:
Yes on `ExecutorStep`, not `DurableTaskStep`, sorry if I was unclear. And yes there could be a similar fix for https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/d9f665d27c269f8294e3897f54f1f7d1f3d6a8e5/pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/Label.java#L74-L80 as well, though I seem to recall Declarative snippet generation being more generally broken (maybe false memory).

jglick commented:
https://github.com/jenkinsci/workflow-durable-task-step-plugin/blob/cc78e242dac43e333be8969952721c2381827f6b/src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStep.java#L98-L114 specifically

Zastai commented:
I have changes in progress for that and the one other use of the old validation method (other than my own plugin), in `matrix-project-plugin` (although that is still based on 2.60.3 so I don’t see that getting merged).

Zastai commented:
Dropped my attempts to change `matrix-project-plugin`; it would first need to be updated to relatively recent dependency versions.

jglick commented:
> Dropped my attempts to change `matrix-project-plugin`; it would first need to be updated to relatively recent dependency versions.

Check https://github.com/jenkinsci/matrix-project-plugin/pull/62

Zastai commented:
OK, three plugins now have reference implementations of the changed API; main PR comment updated.

Main open item from my end is whether or not I should move the i18n messages (to `Messages.LabelExpression_xxx` from `Messages.AbstractProject_xxx`). Messages is no-external-use, so there shouldn't be outside code referencing them, but I want to be sure it's OK.

oleg-nenashev commented on core/src/main/java/hudson/model/labels/LabelExpression.java:
```diff
@@ -210,4 +225,156 @@ public LabelOperatorPrecedence precedence() {
+     * Utility class for taking the current input value and computing a list of potential terms to match against the
+     * list of defined labels.
+     */
+    static class AutoCompleteSeeder {
```
Note that many plugins have copy-pasted the code: https://github.com/search?q=org%3Ajenkinsci+AutoCompleteSeeder&type=Code .
Maybe it;s time to consider making it a public API in `jenkins.model.labels`

Zastai replied:
Makes sense, although it may have been private in part because the auto-completion for labels is still far from perfect - "foo&&b" will pop up completion for the "b" part, but selecting "bar" then replaces the entire expression (mitigated slightly by setting ' ' as delimiter and typing "foo && b").

To me, the fact that many plugins copied the internal implentation just shows how useful exposing an `autoComplete` method is; it does not necessarily make a case for exposing that method's internals.

Having said that, I don't feel all that strongly about _not_ exposing it either.

oleg-nenashev commented on core/src/main/java/hudson/model/labels/LabelExpression.java:
```diff
@@ -210,4 +225,156 @@ public LabelOperatorPrecedence precedence() {
+     * Utility class for taking the current input value and computing a list of potential terms to match against the
+     * list of defined labels.
+     */
+    static class AutoCompleteSeeder {
```
```suggestion
    static class LabelAutoCompleteSeeder {
```

Zastai replied:
Yes, if moving it out of `LabelExpression`, adding the `Label` prefix makes total sense.

oleg-nenashev commented on core/src/main/java/hudson/model/labels/LabelExpression.java:
```diff
@@ -210,4 +225,156 @@ public LabelOperatorPrecedence precedence() {
+     *
+     * @since TODO
+     */
+    public interface LabelValidator extends ExtensionPoint {
```
I would suggest to move it into `jenkins.model.labels` instead of introducing a new nested interface

Zastai replied:
I agree - I initially tried to keep the implementation as close to the original as possible.

oleg-nenashev commented on core/src/main/java/hudson/model/labels/LabelExpression.java:
```diff
@@ -210,4 +225,156 @@ public LabelOperatorPrecedence precedence() {
+     * @return A set of auto-completion candidates.
+     * @since TODO
+     */
+    public static AutoCompletionCandidates autoComplete(@Nullable String label) {
```
```suggestion
    @NonNull
    public static AutoCompletionCandidates autoComplete(@Nullable String label) {
```

Zastai replied:
Yes. After applying the other suggested changes, I'll make another pass to ensure all new/changed API has complete nullability annotations.

oleg-nenashev commented on core/src/main/java/hudson/model/AbstractProject.java:
```diff
@@ -2128,7 +2048,9 @@ public void setCustomWorkspace(String customWorkspace) throws IOException {
      * @since 1.540
+     * @deprecated Use {@link LabelExpression.LabelValidator} instead.
      */
+    @Deprecated
```
This approach makes it difficult to adopt the new API without bumping the core dependency. E.g. I could not update https://plugins.jenkins.io/label-verifier/ . Maybe it makes sense to introduce a temporary bridge method in this class so that checks could be added in the existing extension point.

Does not block it

Zastai replied:
I mean, the implementation for both validate and auto-complete is standalone.
So it _could_ be created as a separate plugin/library, containing just this functionality (wrapped in some new class like `LabelUtil`, not inside `LabelExpression`). Then new core could use that "plugin", and existing plugins could add a dependency on that rather than a newer Jenkins core.

I'm just not sure a fairly minor feature like this warrants that.

oleg-nenashev replied:
It can be done without a plugin and without complex code actually

1. Update LabelValidator

```
public FormValidation checkJob(@NonNull Job<?, ?> job, @NonNull Label label) {
    return FormValidation.ok()
}
```

2. In LabelExpression#validate() add invocation of this method and fail if not OK

3. In a plugin implement a method without adding explicit `@Override` annotation. Thanks to Java, it will still be an override which does not require core bump

Zastai replied:
Ah you meant for the LabelValidator, so for plugins wanting to add validation processing, not plugins wanting to consume the new API.
Sure, I can do that - but I think I'll make the default implementation use `Item`, and call through to the plain `check()` when the passed item is an `AbstractProject`.

oleg-nenashev reviewed (approved):
It looks good in principle, hence I am approving it.
At the same time it could be possible to enhance how API is provided and to simplify adoption of the API in the code

Zastai committed:
Move LabelValidator to the top level.

Zastai committed:
Move LabelExpression.AutoCompleteSeeder to top level as LabelAutoCompleteSeeder.

This also adds missing nullability annotations.

Zastai committed:
Moved the i18n messages.

These are now under LabelExpression.xxx instead of AbstractProject.xxx.

For the translations, properties were either renamed in-place or moved, depending on whether the file seemed to be sorted alphabetically by property name.

Zastai commented:
Suggestions applied; messages moved (no message changes, except that a stray trailing '\' was dropped in the Japanese file).

Zastai commented on core/src/main/java/hudson/model/labels/LabelAutoCompleteSeeder.java:
```diff
@@ -0,0 +1,49 @@
+
+/**
+ * Utility class for taking the current input value and computing a list of potential terms to match against the
+ * list of defined labels.
```
Yes, missed that.

Zastai committed:
Add JavaDoc.

Zastai committed:
Add a checkItem() to the "old" LabelValidator.

By default, this calls the regular check() if the item is an AbstractProject;
otherwise, it returns OK.
This allows a plugin to implement an override for that method in order to have its
validator applied to Jobs too, without needing to bump their Jenkins dependency in
order to get the new LabelValidator.

As a result, LabelExpression.validate() now always uses the "old" LabelValidators,
calling this new method.

Zastai committed:
Merge branch 'master' into JENKINS-26097

jglick commented on core/src/main/java/hudson/model/labels/LabelAutoCompleteSeeder.java:
```diff
@@ -0,0 +1,61 @@
+package hudson.model.labels;
```
I think we prefer to use `jenkins.*` packages for new classes right @oleg-nenashev / @daniel-beck ?

Zastai replied:
I can move them, no problem.

oleg-nenashev replied:
Yes, we should keep new API in "jenkins" if possible. Hudson is still a subject to potential trademark disputes

jglick commented on core/src/main/java/hudson/model/labels/LabelExpression.java:
```diff
@@ -340,27 +278,27 @@ public static FormValidation validate(@Nullable String expression, @CheckForNull
-                    }
+            // Use the project-oriented validators (including any that might implement checkItem to support non-Project
+            // items too).
+            // FIXME: Perhaps these should aggregate their errors/warnings?
```
Unnecessary IMO; `AbstractProject.LabelValidator` should just be deprecated.

Zastai replied:
Without applying `AbstractProject.LabelValidator` Instances, existing plugins implementing that extension point would stop being applied (and there would be no way for them to support pipeline jobs without changing their core dependency).

Zastai committed:
Move new classes to jenkins.model.labels.

Moved the test class over as well.

This also makes the constructor and getSeeds() method of
LabelAutoCompleteSeeder public.

oleg-nenashev reviewed (approved):
Looks good to me after the patches, thanks a lot @Zastai !

jglick commented on core/src/main/java/jenkins/model/labels/LabelAutoCompleteSeeder.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Utility class for taking the current input value and computing a list of potential terms to match against the
+ * list of defined labels.
+ *
+ * @since TODO
+ */
```
```suggestion
 */
@Restricted(NoExternalUse.class)
```

since I do not see it being used in any of the downstream PRs and I would expect it to be called only from `LabelExpression.autoComplete`.

jglick replied:
(plus imports of course)

Zastai replied:
I'm confused. First I get asked to move it out of the class as a public type so it can be reused, and now you want me to flag it to prevent use.

jglick replied:
https://github.com/jenkinsci/jenkins/pull/4774#discussion_r440044574 for reference. Indeed there are a few plugins which have copied this class. The question is whether they actually needed to, or if they could call `LabelExpression.autoComplete` instead. If there is a legitimate need to refer to `LabelAutocompleteSeeder` directly, it could be unrestricted if and when there is a downstream PR demonstrating its proper use.

jglick replied:
Sorry for the mixup, just trying to avoid introducing new public APIs without a clear need.

Zastai replied:
I don't really see why anyone would need to use it. I'd personally be in favour to move it back as a private nested implementation detail class.

jglick replied:
Assuming that a quick review of the existing copies suggests that they could use the simple `autoComplete` method, then yes a private nested implementation class would be fine. I see a usage from test sources though it was not clear if that was necessary or not. (Note that `@Restricted` is ignored in `src/test/java/**/*.java`. It is designed only to prevent improper use of types in _production_ code.)

Zastai replied:
I'll do an org search and check, but I very much suspect all cases copied the seeder because they also copied the autocomplete method, which used the seeder.
As for the test, I think the nested class was package-private to make those tests work. I'm not super convinced that unit tests for the seeder make sense though, given that it's just an implementation detail. But it was there, so I kept it, and will continue to keep it until asked otherwise.

Zastai replied:
Each of the cases found by the org search from https://github.com/jenkinsci/jenkins/pull/4774#discussion_r440044574 is a plugin that copied the doAutoCompleteXXX() method, and also copied the seeder - some as a nested class, some as a separate one. Think I'll just apply the `@Restricted` as suggested - it's less work than moving it back and adapting the unit test again too.

Zastai replied:
Done

jglick reviewed (changes requested):
Requesting API surface reduction. Otherwise looks good.

Zastai commented on core/src/main/resources/hudson/model/Messages.properties:
```diff
@@ -195,6 +188,11 @@ Job.you_must_use_the_save_button_if_you_wish=You must use the Save button if you
+LabelExpression.InvalidBooleanExpression=Invalid boolean expression: {0}
+LabelExpression.LabelLink=<a href="{0}{2}">Label {1}</a> is serviced by {3,choice,0#no nodes|1#1 node|1<{3} nodes}{4,choice,0#|1# and 1 cloud|1< and {4} clouds}. \
+  Permissions or other restrictions provided by plugins may prevent this job from running on those nodes.
+LabelExpression.NoMatch=There''s no agent/cloud that matches this assignment
```
Sure. But it feels like that warrants a separate PR specifically to deal with apostrophes in messages - it seems unlikely that this is the only case. That PR could then deal with the desired approach for Unicodifying punctuation (e.g using "proper" quotation marks too, if they appear in messages).

jglick replied:
Yes, just suggesting it here since you are touching the line of code anyway.

Zastai replied:
Is prose preferred over terseness btw? This specific message sounds like "No agent/cloud matches this label expression" would be fine (also dropping the "assignment" term, which I think is outdated and/or not applicable in all contexts.

jglick replied:
> No agent/cloud matches this label expression

Zastai replied:
Done. I also reworded the other label validation messages to better match each other, and to avoid mentioning "jobs" when the validation may well be happening completely outside of any job context.

Zastai committed:
Merge branch 'master' into JENKINS-26097

Zastai committed:
Tweak messages.

This avoids referring to a job (because the label validation can be
applied completely outside of any job context, like in tool installers)
or an assignment.

Also aligned terms ("is serviced by" -> "matches").

Zastai committed:
Aggregate label validation problems.

This still discards all OKs, but instead of stopping at and returning
the first warning/error reported, all warnings/errors are returned as
an aggregated result.

The JavaDoc for the relevant messages was extended to describe this
behaviour.

Zastai committed:
Flag LabelAutoCompleteSeeder as NoExternalUse.

Zastai commented on core/src/main/java/jenkins/model/labels/LabelValidator.java:
```diff
@@ -0,0 +1,27 @@
+     *
+     * @param item  The context item to be restricted by the label.
+     * @param label The label that the job wants to restrict itself to.
+     * @return The validation result.
```
Done. I also adjusted the use of these validators so that all warnings and errors they report are aggregated and shown.

jglick commented on core/src/main/java/hudson/model/labels/LabelExpression.java:
```diff
@@ -288,28 +289,35 @@ public static FormValidation validate(@Nullable String expression, @CheckForNull
-                if (!FormValidation.Kind.OK.equals(result.kind)) {
-                    return result;
+                if (FormValidation.Kind.OK.equals(result.kind)) {
+                    continue;
```
BTW I think if https://github.com/jenkinsci/jenkins/blob/339f9b930e13c78c545a2409e016127e36520c87/core/src/main/java/hudson/util/FormValidation.java#L227-L238 were refined slightly, to ignore all occurrences of `FormValidation.OK` in its input, then the API and this impl could be simplified a bit while actually handling `ok(String)` from validators. Not necessary in this PR, just something I noticed while looking at `aggregate`.

Zastai replied:
Yes I looked there too. But I'm not sure it would help to get bulleted list entries for "Label is valid" (assuming a validator might put that in the OK text) among the erroneous ones.

jglick replied:
Right—`aggregate` would need to be made smarter, perhaps:

* if no arguments, or all `FormValidation.OK`, return `FormValidation.OK`
* if only one non-`FormValidation.OK` argument, return that as is
* else return worst status with bulleted list of messages

oleg-nenashev reviewed (approved):
Looks good to me, thanks for the adjustments in the pull request!

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process
